tools/congestion-model/src/strategy/soft_backpressure.rs

```diff
@@ -0,0 +1,210 @@
+use crate::model::ChunkExecutionContext;
+use crate::strategy::QueueFactory;
+use crate::{GGas, ShardId, GAS_LIMIT, TGAS, TX_GAS_LIMIT};
+use std::collections::{BTreeMap, HashMap};
+
+#[derive(Default)]
+pub struct SoftBackpressure {
+    shard_id: Option<ShardId>,
+    all_shards: Vec<ShardId>,
+}
+
+#[derive(Default, Clone, Debug)]
+struct ShardState {
+    orig_rec_txn_input_gas: HashMap<(ShardId, ShardId), GGas>,
```

Comment:
I'm also not very fond of the (shard x shard) approach, as beyond a few shards this explodes quickly, and limiting a high dimensional input is unlikely to be effective. For example if we limit (56, 80) because 80 is overloaded and we saw a lot of traffic coming from (56, 80), next block we may see a (61, 80). But if we limit all of (x, 80) then because we make an independent decision on x (source shard), we'd have to apply a very low limit and end up not being able to take in any transactions at all. I'm not sure what to do.